horovod/tensorflow/__init__.py

```diff
@@ -343,26 +344,47 @@ def _make_allreduce_grads_fn(name, device_dense, device_sparse,
                 grads = [tf.convert_to_tensor(grad)
                          if grad is not None and isinstance(grad, tf.IndexedSlices)
                          else grad for grad in grads]
 
-            if num_groups > 0:
-                grads_clean = [grad for grad in grads if grad is not None]
-                grads_split = split_list(grads_clean, num_groups)
-
-                reduce_ops = []
+            if groups is not None:
+                if isinstance(groups, list):
+                    var_name2grad = {}
+                    for i in range(len(vars)):
+                        var = vars[i]
+                        grad = grads[i]
+                        if grad is not None:
+                            var_name2grad[var.name] = (i, grad)
+                    grads_split = []
+                    for group in groups:
+                        grad_group = []
+                        for var in group:
+                            if var.name in var_name2grad:
+                                grad_group.append(var_name2grad[var.name])
+                                del var_name2grad[var.name]
+                        grads_split.append(grad_group)
+                    for _, grad in var_name2grad.items():
+                        grads_split.append([grad])
+                elif groups > 0:
+                    grads_clean = [(i, grad) for i, grad in enumerate(grads) if grad is not None]
+                    grads_split = split_list(grads_clean, groups)
+
+                reduce_ops = [None] * len(vars)
                 for group in grads_split:
-                     reduce_ops += _grouped_allreduce_cond(group,
-                                                          device_dense=device_dense,
-                                                          device_sparse=device_sparse,
-                                                          compression=compression,
-                                                          op=op,
-                                                          prescale_factor=prescale_factor,
-                                                          postscale_factor=postscale_factor)
+                    index_group, grad_group = [list(t) for t in zip(*group)]
+                    reduce_ops_group = _grouped_allreduce_cond(grad_group,
+                                                               device_dense=device_dense,
+                                                               device_sparse=device_sparse,
+                                                               compression=compression,
+                                                               op=op,
+                                                               prescale_factor=prescale_factor,
+                                                               postscale_factor=postscale_factor)
+                    for i in range(len(index_group)):
+                        reduce_ops[index_group[i]] = reduce_ops_group[i]
```

Comment:
The `reduce_ops` list does not need to be submitted to TF in a sorted order. The TF scheduler should handle operation ordering automatically per worker and the Horovod runtime can handle out of order allreduce requests between workers.

Reply:
@romerojosh 
We will zip the variables and gradients with `list(zip(avg_grads, vars))` in `compute_gradients`, where `avg_grads` is the return value of `_allreduce_grads`. Therefore, we need to make sure the reduced grad is in the same order of the input vars.

Reply:
Ok, now I see the issue you mentioned about missing the `None` grad tensors in the returned list in the groups path. Thanks for catching that!